lib/common/common/src/mmap_hashmap.rs

```diff
@@ -0,0 +1,233 @@
+
+        // == First pass ==
+
+        let mut file_size = 0;
+        // 1. Header
+        file_size += size_of::<Header>();
+
+        // 2. PHF
+        file_size += phf.write_bytes();
+
+        // 3. Padding
+        let padding_len = (PADDING_SIZE - (file_size % PADDING_SIZE)) % PADDING_SIZE;
+        file_size += padding_len;
+
+        // 4. Buckets
+        let buckets_pos = file_size;
+        file_size += map.len() * size_of::<u64>();
+
+        // 5. Data
+        let mut buckets = vec![0u64; map.len()];
+        let mut last_bucket = 0;
+        for (k, v) in map.iter() {
+            buckets[phf.get(k).expect("Key not found in phf") as usize] = last_bucket as u64;
+            last_bucket += entry_size(k, v);
+        }
+        file_size += last_bucket;
+        _ = file_size;
+
+        // == Second pass ==
+
+        let mut bufw = io::BufWriter::new(File::create(path)?);
+
+        // 1. Header
+        let header = Header {
+            buckets_pos: buckets_pos as u64,
+            buckets_count: map.len() as u64,
+        };
+        bufw.write_all(header.as_bytes())?;
+
+        // 2. PHF
+        phf.write(&mut bufw)?;
+
+        // 3. Padding
+        bufw.write_all(zeroes(padding_len))?;
+
+        // 4. Buckets
+        bufw.write_all(buckets.as_bytes())?;
+
+        // 5. Data
+        let mut pos = 0;
+        for (k, v) in map.iter() {
+            buckets.push(pos as u64);
+            pos += entry_size(k, v);
+
+            // Lengths
+            bufw.write_all((k.len() as u64).as_bytes())?;
+            bufw.write_all((v.len() as u64).as_bytes())?;
+
+            // Key
+            bufw.write_all(k.as_bytes())?;
+
+            bufw.write_all(zeroes(k.len().next_multiple_of(4) - k.len()))?;
+
+            // Values
+            for i in v.iter() {
+                bufw.write_all(i.as_bytes())?;
+            }
+        }
+        bufw.flush()?;
+
+        Ok(())
+    }
+
+    /// Load the hash map from file.
+    pub fn open(path: &Path) -> io::Result<Self> {
+        let file = File::open(path)?;
+
+        // SAFETY: Assume other processes do not modify the file.
+        // See https://docs.rs/memmap2/latest/memmap2/struct.Mmap.html#safety
+        let mmap = unsafe { Mmap::map(&file)? };
+
+        let header = Header::read_from_prefix(mmap.as_ref()).ok_or(io::ErrorKind::InvalidData)?;
+
+        let phf = Function::read(&mut Cursor::new(
+            &mmap
+                .get(size_of::<Header>()..header.buckets_pos as usize)
+                .ok_or(io::ErrorKind::InvalidData)?,
+        ))?;
+
+        Ok(MmapHashMap { mmap, header, phf })
+    }
+
+    /// Get the values associated with `key`. It will return `None` in the following cases:
+    /// - The key is not present in this hash map.
+    /// - The data is corrupted, e.g. invalid UTF-8 or wrong alignment. (should not happen during
+    ///   normal operation)
+    pub fn get(&self, key: &str) -> Option<&[u32]> {
```

Comment:
Corrupted storage is not the same as unexist key. I propose to add error handling and return here `OperationResult<Option<&[u32]>>`

Reply:
Thanks!